src/Http/Routing/src/EndpointSelectorContext.cs

```diff
@@ -3,68 +3,52 @@
 
 using System;
 using Microsoft.AspNetCore.Http;
-using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Http.Endpoints;
 
 namespace Microsoft.AspNetCore.Routing
 {
-    public sealed class EndpointSelectorContext : IEndpointFeature, IRouteValuesFeature, IRoutingFeature
+    public struct EndpointSelectorContext
     {
-        private RouteData _routeData;
-        private RouteValueDictionary _routeValues;
+        private HttpContext _httpContext;
 
-        /// <summary>
-        /// Gets or sets the selected <see cref="Http.Endpoint"/> for the current
-        /// request.
-        /// </summary>
-        public Endpoint Endpoint { get; set; }
+        public EndpointSelectorContext(HttpContext httpContext)
+        {
+            _httpContext = httpContext;
+        }
 
         /// <summary>
-        /// Gets or sets the <see cref="RouteValueDictionary"/> associated with the currrent
+        /// Gets or sets the selected <see cref="Http.Endpoint"/> for the current
         /// request.
         /// </summary>
-        public RouteValueDictionary RouteValues
+        public Endpoint Endpoint
         {
-            get => _routeValues ?? (_routeValues = new RouteValueDictionary());
+            get
+            {
+                return _httpContext.GetEndpoint();
+            }
             set
             {
-                _routeValues = value;
-
-                // RouteData will be created next get with new Values
-                _routeData = null;
+                if (value != null)
```

Comment:
This should still mutate the context. It needs to be possible to set it to null.